analyses/notebooks/rl_intro_notebook.ipynb

```diff
@@ -0,0 +1,2171 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "id": "OEfL2KEhJKWz"
+   },
+   "outputs": [],
+   "source": [
+    "class Q_table():\n",
+    "    \"\"\"\n",
+    "    Class representing the Q_table\n",
+    "\n",
+    "    Attributes:\n",
+    "        table: the current Q table\n",
+    "        updated_table: the version of the Q table during updating (training)\n",
+    "    \"\"\"\n",
+    "    def __init__(self, par: Parameters):\n",
+    "        self.table = {}\n",
+    "        self.updated_table = {}\n",
+    "        print('Q_table class initialized...')\n",
+    "\n",
+    "    def update_q_table(self, sample: np.ndarray) -> None:\n",
+    "        \"\"\"\n",
+    "        Updates/trains the q_table values\n",
+    "\n",
+    "        Args:\n",
+    "            sample: the collection of observations (state, action, reward, next_state) gathered by the agent during the iteration\n",
+    "        \"\"\"\n",
+    "        # retrieve the state, action, reward, and next state from the sample\n",
+    "        obs = [(tuple(obs[par.index_state:par.index_action]),\n",
+    "                tuple(obs[par.index_action:par.index_reward]), \n",
+    "                int(obs[par.index_reward:par.index_next_state]),\n",
+    "                tuple(obs[par.index_next_state:]))\n",
+    "               for obs in sample]\n",
+    "\n",
+    "        # For each observation...\n",
+    "        for (state, action, reward, next_state) in obs:\n",
+    "\n",
+    "            # If the state is encountered for the first time, add it to the table\n",
+    "            if state not in self.updated_table.keys(): # the state is not yet encountered\n",
+    "                self.updated_table[state] = {action: reward}\n",
+    "      \n",
+    "            # if the state-action combination is not yet encountered (but the state has), add it to the table\n",
+    "            elif action not in self.updated_table[state].keys():\n",
+    "                self.updated_table[state][action] = reward\n",
+    "      \n",
+    "            # if the state-action pair has been encountered before, update the table\n",
+    "            # value using the Bellmann equation\n",
+    "            else:\n",
+    "                if next_state in self.table.keys():\n",
+    "                    future_exp_reward = self.table[next_state][self.get_best_action(next_state)]\n",
+    "                else:\n",
+    "                    future_exp_reward = 0\n",
+    "                # future_exp_reward = par.gamma * self.table[state][self.get_best_action(next_state)] if state in self.table.keys() else 0\n",
+    "                self.updated_table[state][action] = reward + future_exp_reward\n",
+    "\n",
+    "        # Overwrite the actual Q-table with the updated values\n",
+    "        self.table = deepcopy(self.updated_table)\n",
+    "\n",
+    "    def get_best_action(self, state: np.ndarray) -> np.ndarray:\n",
```

Comment:
does get_best_action now fail if the state has not been visited before? Is that the desired result?